### PR TITLE
Test ekuiper subscribe to EdgeX raw events without using ASC by default

### PR DESCRIPTION
**Pull request by MonicaisHer**
todo:
- [x] Rebase after https://github.com/canonical/edgex-snap-testing/pull/124 has been merged
- [x] Re-run test after https://github.com/canonical/edgex-ekuiper-snap/pull/34 PR has been merged

### CODE REVIEW

File: test/suites/ekuiper/streams_rules_test.go [hunk 1]
```diff
@@ -58,10 +67,39 @@ func TestStreamsAndRules(t *testing.T) {
 	utils.TestDeviceVirtualReading(t)
 
 	t.Run("check rule_log", func(t *testing.T) {
-		utils.Exec(t, `edgex-ekuiper.kuiper-cli getstatus rule rule_log`)
+		//check logs for the record of expected log
+		start := time.Now()
+		time.Sleep(1 * time.Second)
+		logs := utils.SnapLogs(t, start, ekuiperSnap)
```
Comment (farshidtz) [inline test/suites/ekuiper/streams_rules_test.go:73]: Not sure what is the purpose of this. Sleep one second and query logs since a second ago?

Comment (MonicaisHer) [reply test/suites/ekuiper/streams_rules_test.go:73]: It is looking for the sink_log's result in the log. It worked locally. I am wondering if maybe the duration needs to be extended.

**Comment by farshidtz** [comment]
Something isn't right with the tests as they hang [occasionally](https://github.com/canonical/edgex-snap-testing/actions/runs/3297518699/jobs/5471184237#step:5:123):
```
Mon, 24 Oct 2022 14:02:06 GMT
    readings.go:45: waiting for device-virtual produce readings, current retry count: 10/60
Mon, 24 Oct 2022 14:02:06 GMT
    readings.go:48: device-virtual is producing readings now, readings queried from core-data
Mon, 24 Oct 2022 14:02:06 GMT
=== RUN   TestStreamsAndRules/check_rule_log
Mon, 24 Oct 2022 14:02:07 GMT
    exec.go:19: [exec] sudo journalctl --since "2022-10-24 14:02:06" --no-pager | grep "edgex-ekuiper"|| true
Mon, 24 Oct 2022 14:25:27 GMT
Error: The operation was canceled.
```
This happened once when I ran locally too.

Let's put this on hold until for now.

File: test/suites/ekuiper/streams_rules_test.go [hunk 2]
```diff
@@ -55,13 +64,42 @@ func TestStreamsAndRules(t *testing.T) {
...
 	utils.TestDeviceVirtualReading(t)
 
 	t.Run("check rule_log", func(t *testing.T) {
-		utils.Exec(t, `edgex-ekuiper.kuiper-cli getstatus rule rule_log`)
+		//check logs for the record of expected log
+		time.Sleep(1 * time.Second)
+		logs := utils.SnapLogs(t, start, ekuiperSnap)
+		expectLog := "sink result for rule rule_log"
+
+		require.True(t, strings.Contains(logs, expectLog))
```
[the remark this answers is not in the document]
Comment (MonicaisHer) [reply test/suites/ekuiper/streams_rules_test.go:76]: The output of ekuiper CLI is hard to validate. Here is an example:
```
# bin/kuiper getstatus rule rule1
Connecting to 127.0.0.1:20498...
{
  "source_demo_0_records_in_total": 29,
  "source_demo_0_records_out_total": 29,
  "source_demo_0_exceptions_total": 0,
  "source_demo_0_process_latency_ms": 0,
  "source_demo_0_buffer_length": 0,
  "source_demo_0_last_invocation": "2020-04-17T10:30:09.294337",
  "op_filter_0_records_in_total": 29,
  "op_filter_0_records_out_total": 21,
  "op_filter_0_exceptions_total": 0,
  "op_filter_0_process_latency_ms": 0,
  "op_filter_0_buffer_length": 0,
  "op_filter_0_last_invocation": "2020-04-17T10:30:09.294362",
  "op_project_0_records_in_total": 21,
  "op_project_0_records_out_total": 21,
  "op_project_0_exceptions_total": 0,
  "op_project_0_process_latency_ms": 0,
  "op_project_0_buffer_length": 0,
  "op_project_0_last_invocation": "2020-04-17T10:30:09.294382",
  "sink_mqtt_0_0_records_in_total": 21,
  "sink_mqtt_0_0_records_out_total": 21,
  "sink_mqtt_0_0_exceptions_total": 0,
  "sink_mqtt_0_0_process_latency_ms": 0,
  "sink_mqtt_0_0_buffer_length": 1,
  "sink_mqtt_0_0_last_invocation": "2020-04-17T10:30:09.294423"
}
```

**Comment by MonicaisHer** [comment]
I think the rule_edgex_message_bus causes the hang-on issue, because it republish readings back to the message bus which leads to exponential readings in log

**Comment by MonicaisHer** [comment]
I ran the test 20 times and all passed:
```
$ go test -p 1 ./test/suites/ekuiper --count=20
ok      edgex-snap-testing/test/suites/ekuiper  638.843s
```

**Comment by farshidtz** [comment]
> Also, there is about 10 -15 s latency between readings produced by the device-virtual and received by ekuiper rule.

It is better to check the expected value periodically for a predefined period. It may take even longer on a slow CI system. In other cases it may take less time.

File: test/suites/ekuiper/streams_rules_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MonicaisHer) [reply test/suites/ekuiper/streams_rules_test.go]: Make sense. Updated.

File: test/suites/ekuiper/streams_rules_test.go [hunk 3]
```diff
@@ -58,10 +71,86 @@ func TestStreamsAndRules(t *testing.T) {
...
+
+			startIndex := strings.Index(stdout, "{")
+
+			if startIndex < 0 {
+				t.Fatal("Error getting status of rule rule_log in JSON format")
+			} else {
+				body = stdout[startIndex:]
+			}
+
+			err = json.Unmarshal([]byte(body), &ruleStatus)
```
Comment (farshidtz) [inline test/suites/ekuiper/streams_rules_test.go:93]: I now understand why you said it was difficult. Why don't you just call the REST API with http.Get?

```
$ curl http://localhost:59720/rules/rule_log/status
{
  "source_stream1_0_records_in_total": 0,
  "source_stream1_0_records_out_total": 0,
  "source_stream1_0_exceptions_total": 0,
  "source_stream1_0_process_latency_us": 0,
  "source_stream1_0_buffer_length": 0,
  "source_stream1_0_last_invocation": 0,
  "op_2_filter_0_records_in_total": 0,
  "op_2_filter_0_records_out_total": 0,
  "op_2_filter_0_exceptions_total": 0,
  "op_2_filter_0_process_latency_us": 0,
  "op_2_filter_0_buffer_length": 0,
  "op_2_filter_0_last_invocation": 0,
  "op_3_project_0_records_in_total": 0,
  "op_3_project_0_records_out_total": 0,
  "op_3_project_0_exceptions_total": 0,
  "op_3_project_0_process_latency_us": 0,
  "op_3_project_0_buffer_length": 0,
  "op_3_project_0_last_invocation": 0,
  "sink_log_0_0_records_in_total": 0,
  "sink_log_0_0_records_out_total": 0,
  "sink_log_0_0_exceptions_total": 0,
  "sink_log_0_0_process_latency_us": 0,
  "sink_log_0_0_buffer_length": 0,
  "sink_log_0_0_last_invocation": 0
}
```

File: test/suites/ekuiper/streams_rules_test.go [hunk 4]
```diff
@@ -2,21 +2,34 @@ package test
...
 	"testing"
+	"time"
 )
 
+type Reading struct {
+	TotalCount int `json:"totalCount"`
+}
+
+type RuleStatus struct {
+	SouceCount int `json:"source_stream1_0_records_out_total"`
```
Comment (farshidtz) [inline test/suites/ekuiper/streams_rules_test.go:19]: SourceCount

File: test/suites/ekuiper/streams_rules_test.go [hunk 5]
```diff
@@ -58,10 +71,86 @@ func TestStreamsAndRules(t *testing.T) {
...
+		} else {
+			body = stdout[startIndex:]
+		}
+
+		err = json.Unmarshal([]byte(body), &ruleStatus)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		require.Greaterf(t, ruleStatus.LogCount, 0, "No readings have been published to log by ekuiper")
```
Comment (farshidtz) [inline test/suites/ekuiper/streams_rules_test.go:129]: Doesn't a LogCount of greater than 0 imply the source count of greater than 0? I think checking LogCount in the above loop may be sufficient.

File: test/suites/ekuiper/streams_rules_test.go [hunk 6]
```diff
@@ -58,10 +71,86 @@ func TestStreamsAndRules(t *testing.T) {
...
+			t.Fatal(err)
+		}
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = json.Unmarshal(body, &reading)
```
Comment (farshidtz) [inline test/suites/ekuiper/streams_rules_test.go:149]: Read and decode together, if you want to reduce the number of lines: `json.NewDecoder(resp.Body).Decode(&reading)`

File: test/suites/ekuiper/streams_rules_test.go (hunk 3 above)
Comment (MonicaisHer) [reply test/suites/ekuiper/streams_rules_test.go:93]: Oh, I didn't aware that the results are different between kuiper CLI and REST API. Updated!

Comment (farshidtz) [reply test/suites/ekuiper/streams_rules_test.go:93]: I think this is really a bug on the CLI. Only the JSON should be printed to the standard output and everything else should go to standard error to allow capturing the payload.

Comment (MonicaisHer) [reply test/suites/ekuiper/streams_rules_test.go:93]: Agree. I think `Connecting to 127.0.0.1:20498...` should not be in stdout or stderr at all. maybe it should be in debugging log.
Would you like to open an issue in ekuiper repo?

**Comment by MonicaisHer** [comment]
I ran the test 20 times and all passed:
```
$ go test -p 1 ./test/suites/ekuiper --count=20
ok      edgex-snap-testing/test/suites/ekuiper  522.032s
```

**Review by farshidtz — APPROVED** [review]
Looks good. Thank @MonicaisHer for going through the back and forth reviews.

**Pull request merged**